File: 2023/day7/main.py
from typing import List, NamedTuple, Dict, Callable, Tuple
from collections import Counter
from functools import partial


class Hand(NamedTuple):
    cards: List[str]
    bid: int


CARD_VALUES = {str(i): i for i in range(2, 10)}
CARD_VALUES.update(
    {
        "T": 10,
        "J": 11,
        "Q": 12,
        "K": 13,
        "A": 14,
    }
)

CARD_VALUES2 = CARD_VALUES
CARD_VALUES2["J"] = 1


def read_input(file: str) -> List[str]:
    with open(file, mode="r") as f:
        out = [x.strip() for x in f.readlines()]
    return out


def parse_line(line: str) -> Hand:
    cards, bid = line.split(" ")
    cards = [c for c in cards]
    bid = int(bid)
    return Hand(cards, bid)


def hand_type(hand: Hand) -> int:
    counts = Counter(hand.cards).most_common(2)
    first = counts[0][1]
    if first == 5:
        second = 0
    else:
        second = counts[1][1]
    if first == 5:
        return 6
    elif first == 4:
        return 5
    elif first == 3 and second == 2:
        return 4
    elif first == 3:
        return 3
    elif first == 2 and second == 2:
        return 2
    elif first == 2:
        return 1
    else:
        return 0


def hand_type2(hand: Hand) -> int:
    n_jokers = len([c for c in hand.cards if c == "J"])
    if n_jokers == 5:
        return 6
    counts = Counter(c for c in hand.cards if c != "J").most_common(2)
    first = counts[0][1]
    if len(counts) == 1:
        second = 0
    else:
        second = counts[1][1]
    if first + n_jokers == 5:
        return 6
    elif first + n_jokers == 4:
        return 5
    elif first + second + n_jokers == 5:
        return 4
    elif first + n_jokers == 3:
        return 3
    elif (
        n_jokers >= 2 or (n_jokers == 1 and first == 2) or (first == 2 and second == 2)
    ):
        return 2
    elif n_jokers == 1 or first == 2:
        return 1
    else:
        return 0


def hand_order(
    hand: Hand, hand_type: Callable[Hand, int], card_values: Dict[str, int]
) -> Tuple[int, List[int]]:
    return (hand_type(hand), [card_values[c] for c in hand.cards])


input = read_input("input.txt")
hands = [parse_line(x) for x in input]

sort1 = partial(hand_order, hand_type=hand_type, card_values=CARD_VALUES)
hands.sort(key=sort1)
part1 = sum((i + 1) * h.bid for (i, h) in enumerate(hands))
print(f"Part 1: {part1}")

sort2 = partial(hand_order, hand_type=hand_type2, card_values=CARD_VALUES2)
hands.sort(key=sort2)
part2 = sum((i + 1) * h.bid for (i, h) in enumerate(hands))
print(f"Part 2: {part2}")

File: 2023/day7/src/main.rs
use std::fs::read_to_string;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum Card {
    N2,
    N3,
    N4,
    N5,
    N6,
    N7,
    N8,
    N9,
    T,
    J,
    Q,
    K,
    A,
}

impl TryFrom<char> for Card {
    type Error = &'static str;

    fn try_from(value: char) -> Result<Self, Self::Error> {
        match value {
            '2' => Ok(Card::N2),
            '3' => Ok(Card::N3),
            '4' => Ok(Card::N4),
            '5' => Ok(Card::N5),
            '6' => Ok(Card::N6),
            '7' => Ok(Card::N7),
            '8' => Ok(Card::N8),
            '9' => Ok(Card::N9),
            'T' => Ok(Card::T),
            'J' => Ok(Card::J),
            'Q' => Ok(Card::Q),
            'K' => Ok(Card::K),
            'A' => Ok(Card::A),
            _ => Err("Unable to parse card"),
        }
    }
}

#[derive(Debug, PartialEq)]
struct Hand {
    cards: [Card; 5],
    bid: i32,
}

impl TryFrom<&String> for Hand {
    type Error = &'static str;

    fn try_from(value: &String) -> Result<Self, Self::Error> {
        let cards = TryInto::<[Card; 5]>::try_into(
            value
                .split(' ')
                .nth(0)
                .unwrap()
                .chars()
                .map(|x| Card::try_from(x).unwrap())
                .collect::<Vec<Card>>(),
        );

        let bid = value.split(' ').nth(1).unwrap().parse::<i32>();

        match (cards, bid) {
            (Ok(cards), Ok(bid)) => Ok(Hand { cards, bid }),
            _ => Err("Unable to parse hand"),
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
enum HandType {
    HighCard,
    OnePair,
    TwoPair,
    ThreeOfAKind,
    FullHouse,
    FourOfAKind,
    FiveOfAKind,
}


fn hand_type_jack(hand: &Hand) -> HandType {
    let mut counts = [0; 13];
    for card in hand.cards.iter(){
        counts[*card as usize] += 1;
    }
    counts.sort();
    counts.reverse();
    let most_common = counts.first();
    let second_most_common = counts.get(1);

    match (most_common, second_most_common) {
        (Some(5), _) => HandType::FiveOfAKind,
        (Some(4), _) => HandType::FourOfAKind,
        (Some(3), Some(2)) => HandType::FullHouse,
        (Some(3), _) => HandType::ThreeOfAKind,
        (Some(2), Some(2)) => HandType::TwoPair,
        (Some(2), _) => HandType::OnePair,
        _ => HandType::HighCard,
    }
}

fn hand_value_jack(hand: &Hand) -> u64 {
    // all enums < 16 so can store with 4 bits
    let mut value = hand_type_jack(hand) as u64;
    for card in hand.cards {
        value = (value << 4) | (card as u64);
    }
    value
}

fn hand_type_joker(hand: &Hand) -> HandType {
    let mut counts = [0; 13];
    for card in hand.cards.iter(){
        counts[*card as usize] += 1;
    }
    let n_jokers = counts[Card::J as usize];
    counts[Card::J as usize] = 0;
    counts.sort();
    counts.reverse();
    let most_common = counts[0];
    let second_most_common = counts[1];
    if n_jokers + most_common == 5 {
        HandType::FiveOfAKind
    } else if n_jokers + most_common == 4 {
         HandType::FourOfAKind
    } else if n_jokers + most_common + second_most_common == 5 {
         HandType::FullHouse
    } else if n_jokers + most_common == 3 {
        HandType::ThreeOfAKind
    } else if n_jokers + most_common + second_most_common == 4 {
        HandType::TwoPair
    } else if n_jokers == 1 || most_common == 2 {
        HandType::OnePair
    } else {
        HandType::HighCard
    }
}

fn hand_value_joker(hand: &Hand) -> u64 {
    // all enums < 16 so can store with 4 bits
    let mut value = hand_type_joker(hand) as u64;
    for card in hand.cards {
        let card_value = if card == Card::J {1} else {(card as u64) + 2};
        value = (value << 4) | card_value;
    }
    value
}

fn read_input(filename: &str) -> Vec<String> {
    read_to_string(filename)
        .unwrap()
        .lines()
        .map(String::from)
        .collect()
}

fn main() {
    let input = read_input("input.txt");

    let mut hands: Vec<Hand> = input
        .iter()
        .map(|x| Hand::try_from(x).expect("Should be able to parse all hands"))
        .collect();

    hands.sort_by_cached_key(hand_value_jack);
    let part1: i32 = hands
        .iter()
        .enumerate()
        .map(|(i, h)| h.bid * (i as i32 + 1))
        .sum();

    println!("Part 1: {:?}", part1);

    hands.sort_by_cached_key(hand_value_joker);
    let part2: i32 = hands
        .iter()
        .enumerate()
        .map(|(i, h)| h.bid * (i as i32 + 1))
        .sum();


    println!("Part 2: {:?}", part2);

}
